### PR TITLE
Add a type hint for `get_device_handler()` and deal with the fallout.

### PR DESCRIPTION
**Pull request by clokep**
I'm not 100% thrilled with the way this came out, but not having a type hint on properties of `HomeServer` is wildly dangerous IMO.

### CODE REVIEW

**Comment by clokep** [comment]
Hm, looks like some real failures. Will need to investigate!

**Review by DMRobertson — COMMENTED** [review]
Drive-by comment: I like the pattern of having an "updater" object which either makes a change or asks an authoritative worker (in this case, the master) to make it for us.

**Comment by DMRobertson** [comment]
This would have caught a slip-up of mine. LMK if you want a hand getting this over the line.

**Comment by clokep** [comment]
> This would have caught a slip-up of mine. LMK if you want a hand getting this over the line.

I mostly just haven't had time to look back at it. If you'd like to, feel free to push more commits! 🎉

**Comment by clokep** [comment]
I'm pretty sure this will pass everything now. :)

File: synapse/handlers/device.py [hunk 1]
```diff
@@ -320,6 +324,8 @@ async def handle_room_un_partial_stated(self, room_id: str) -> None:
 
 
 class DeviceHandler(DeviceWorkerHandler):
```
Comment (DMRobertson) [inline synapse/handlers/device.py:326]: What's the difference between `DeviceHandler` and `DeviceWorkerHandler` now? Has this changed as a result of your patch here?

File: synapse/module_api/__init__.py [hunk 2]
```diff
@@ -784,6 +786,8 @@ def invalidate_access_token(
     ) -> Generator["defer.Deferred[Any]", Any, None]:
         """Invalidate an access token for a user
 
+        Can only be called from the main process.
```
Comment (DMRobertson) [inline synapse/module_api/__init__.py:789]: Is this a new requirement, or are we just making this explicit?

File: synapse/rest/admin/__init__.py [hunk 3]
```diff
@@ -238,6 +238,10 @@ def register_servlets(hs: "HomeServer", http_server: HttpServer) -> None:
     """
     Register all the admin servlets.
     """
+    # Admin servlets aren't registered on workers.
+    if hs.config.worker.worker_app is not None:
+        return
```
Comment (DMRobertson) [inline synapse/rest/admin/__init__.py:243]: I think a few read-only servlets (ListRoom,Version) would probably be safe to run on workers... but we probably don't take advantage of that?

(Just wondering if there's a user-facing change that we need to call out here)

**Review by DMRobertson — APPROVED** [review]
I think this is sane---though I'm still generally baffled by what does and doesn't get allocated to workers etc

File: synapse/module_api/__init__.py (hunk 2 above)
Comment (clokep) [reply synapse/module_api/__init__.py:789]: Just making it explicit.

File: synapse/handlers/device.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/device.py:339]: I moved `get_dehydrated_device` from `DeviceHandler` to `DeviceWorkerHandler`. The only other change is that their `device_list_updater` are now `DeviceListUpdater` or `DeviceListWorkerUpdater`, respectively. The latter has an overridden method (`user_device_resync`) to call across replication to the main process.

**Comment by clokep** [comment]
@DMRobertson This needed more changes, I'm pretty sure it'll pass this time though.

File: synapse/rest/admin/__init__.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/rest/admin/__init__.py:243]: > We do not register any of these for workers right now...except in unit tests.

Super, thanks for spelling it out for me.

**Review by DMRobertson — APPROVED** [review]
Many thanks again for this, I think this really helps clarify the tangled web we've weaved.

Sytest passed this time. Fingers cross complement will too, modulo flakes!

**Pull request merged**